src/Storages/MergeTree/MergeTreeLazilyReader.cpp

```diff
@@ -0,0 +1,277 @@
+    }
+
+    size_t curr_idx = 0;
+    permutation.resize(rows_size);
+
+    for (auto & part_and_row_offsets : part_to_row_offsets)
+    {
+        auto & row_offsets_with_idx = part_and_row_offsets.second;
+
+        std::sort(row_offsets_with_idx.begin(), row_offsets_with_idx.end(), [](const RowOffsetWithIdx & left,
+                                                                                                 const RowOffsetWithIdx & right)
+        {
+            return left.row_offset < right.row_offset;
+        });
+
+        for (const auto & row_offset_with_idx : row_offsets_with_idx)
+        {
+            permutation[row_offset_with_idx.row_idx] = curr_idx++;
+        }
+    }
+}
+
+void addDummyColumnWithRowCount(Block & block, Columns & res_columns, size_t num_rows)
+{
+    bool has_columns = false;
+    for (const auto & column : res_columns)
+    {
+        if (column)
+        {
+            has_columns = true;
+            break;
+        }
+    }
+
+    if (has_columns)
+        return;
+
+    ColumnWithTypeAndName dummy_column;
+    dummy_column.column = DataTypeUInt8().createColumnConst(num_rows, Field(1));
+    dummy_column.type = std::make_shared<DataTypeUInt8>();
+    dummy_column.name = "....dummy...." + toString(UUIDHelpers::generateV4());
+    block.insert(dummy_column);
+}
+
+MergeTreeLazilyReader::MergeTreeLazilyReader(
+    const Block & header_,
+    const MergeTreeData & storage_,
+    const StorageSnapshotPtr & storage_snapshot_,
+    const LazilyReadInfoPtr & lazily_read_info_,
+    const ContextPtr & context_,
+    const AliasToName & alias_index_)
+    : storage(storage_)
+    , data_parts_info(lazily_read_info_->data_parts_info)
+    , storage_snapshot(storage_snapshot_)
+    , use_uncompressed_cache(context_->getSettingsRef()[Setting::use_uncompressed_cache])
+{
+    NameSet columns_name_set;
+
+    for (const auto & column_name : lazily_read_info_->lazily_read_columns)
+        columns_name_set.insert(column_name.name);
+
+    for (const auto & column : header_)
+    {
+        const auto it = alias_index_.find(column.name);
+        if (it == alias_index_.end())
+            continue; // TODO: check if it's what we want
+
+        const auto & requested_column_name = it->second;
+        if (columns_name_set.contains(requested_column_name))
+        {
+            requested_column_names.emplace_back(requested_column_name);
+            lazy_columns.emplace_back(header_.getByName(column.name));
+        }
+    }
+}
+
+void MergeTreeLazilyReader::readLazyColumns(
+    const MergeTreeReaderSettings & reader_settings,
+    const PartIndexToRowOffsets & part_to_row_offsets,
+    MutableColumns & lazily_read_columns)
+{
+    const auto columns_size = lazily_read_columns.size();
+
+    for (const auto & it : part_to_row_offsets)
+    {
+        auto part_index = it.first;
+        const auto & row_offsets = it.second;
+        MergeTreeData::DataPartPtr data_part = (*data_parts_info)[part_index].data_part;
```

Comment:
What if there is no data part info for such `part_index`?

Reply:
Added exception handling for missing part_index.